src/slurm_cloud_bursting/resume.py

```diff
@@ -0,0 +1,262 @@
+
+        config = ConfigParser()
+        try:
+            config.read_file(open(config_file_path, "r"))
+        except IOError:
+            log.error(f"Cannot read slurm cloud bursting scripts configuration file: {config_file_path}")
+            raise
+
+        self.region = config.get("slurm_cb_config", "region")
+        self.cluster_name = config.get("slurm_cb_config", "cluster_name")
+        self.max_batch_size = int(
+            config.get("slurm_cb_config", "max_batch_size", fallback=self.DEFAULT_MAX_INSTANCES_BATCH_SIZE)
+        )
+        self.update_node_address = config.getboolean("slurm_cb_config", "update_node_address", fallback=True)
+
+        # Configure boto3 to retry 5 times by default
+        self._boto3_config = {"retries": {"max_attempts": self.DEFAULT_MAX_RETRY, "mode": "standard"}}
+        proxy = config.get("slurm_cb_config", "proxy", fallback="NONE")
+        if proxy != "NONE":
+            self._boto3_config["proxies"] = {"https": proxy}
+        self.boto3_config = Config(**self._boto3_config)
+
+        log.info(self.__repr__())
+
+
+@retry(stop_max_attempt_number=3, wait_fixed=5000)
+def _set_nodes_down_and_power_save(node_list):
+    set_nodes_down(node_list, reason="Failure when resuming nodes")
+    set_nodes_power_down(node_list, reason="Failure when resuming nodes")
+
+
+def _handle_failed_nodes(node_list):
+    """
+    Fall back mechanism to handle failure when launching instances.
+
+    When encountering a failure, want slurm to deallocate current nodes,
+    and re-queue job to be run automatically by new nodes.
+    To do this, set node to DOWN, so slurm will automatically re-queue job.
+    Then set node to POWER_DOWN so suspend program will be run.
+    Suspend program needs to properly clean up instances(if any) and set node back to IDLE in all cases.
+
+    If this process is not done explicitly, slurm will wait until ResumeTimeout,
+    then execute this process of setting nodes to DOWN then POWER_DOWN.
+    To save time, should explicitly set nodes to DOWN then POWER_DOWN after encountering failure.
+    """
+    try:
+        log.info("Node %s marked as down and placed into power_down", node_list)
+        _set_nodes_down_and_power_save(node_list)
+    except Exception as e:
+        log.exception("Failed to place nodes %s into down/power_down with exception: %s", node_list, e)
+
+
+def _parse_ec2_instance(instance):
+    """Parse and return EC2 instance info."""
+    return EC2Instance(
+        instance["InstanceId"],
+        private_ip=instance["PrivateIpAddress"],
+        hostname=instance["PrivateDnsName"].split(".")[0],
+    )
+
+
+def _update_slurm_node_addrs(launched_nodes):
+    """Update node information in slurm with info from launched EC2 instance."""
+    for instance, slurm_node in launched_nodes:
+        try:
+            update_nodes(slurm_node, nodeaddr=instance.private_ip, nodehostname=instance.hostname, raise_on_error=True)
```

Comment:
can you remind me why we are not batching these calls? If it's to avoid failing the update on all nodes if individual nodes fail I'm wondering if we can do some optimization such as updating in batch and if it fails falling back to individual node updates. The reason is that I expect Slurm might handle batch updates in a more optimized fashion.

Reply:
We are calling this function in batches

Reply:
Verify how slurm handle failure